src/Interpreters/TableJoin.cpp

```diff
@@ -233,9 +233,15 @@ void TableJoin::addJoinedColumnsAndCorrectNullability(ColumnsWithTypeAndName & c
         if (leftBecomeNullable(col.type))
             col.type = makeNullable(col.type);
 
+    std::unordered_map<String, DataTypePtr> type_map;
```

Comment:
So the type information in the input block here is "wrong" (before conversion), and we have to overwrite it w/ the converted values? But why? Can the input block be made to already have the proper types?

Reply:
I guess, if actions for converting types will be in a separate step, this code won't be needed ...

Reply:
I moved  common type calculation into [TreeRewriter](https://github.com/ClickHouse/ClickHouse/pull/19885/files#diff-03665017349505c5529e5d0eeeeae80081e064267d6035b94cd428baf11856ceR421) and now TableJoin contains information about converted types on earlier stage, so it is possible to change types in [addJoinedColumn](https://github.com/ClickHouse/ClickHouse/pull/19885/files#diff-a8a50243ee6c63a7030a0f87fe3e35677f6097063a807ac0e6f4fd0f5dd3b77eR222) and have correct types in columns_added_by_join